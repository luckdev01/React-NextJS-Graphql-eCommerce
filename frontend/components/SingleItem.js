import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Error from './ErrorMessage';
import SingleItemStyles from './styles/SingleItemStyles';
import Head from 'next/head';

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      largeImage
    }
  }
`;

class SingleItem extends Component {
  render() {
    return (
      <Query query={SINGLE_ITEM_QUERY} variables={{ id: this.props.id }}>
        {({ error, loading, data }) => {
          if (error) return <Error error={error} />;
          if (loading) return <p>Loading...</p>;
          if (!data.item) return <p>No item found for {this.props.id}</p>;
          const item = data.item;
          return (
            <SingleItemStyles>
              {/* Head gets inserted into HTML head by nextjs */}
              <Head>
                <title>Sick Fits | {item.title}</title>
              </Head>
              <img src={item.largeImage} alt={item.title} />
              <div className="details">
                <h2>Viewing {item.title}</h2>
                <p>{item.description}</p>
              </div>
            </SingleItemStyles>
          );
        }}
      </Query>
    );
  }
}

export default SingleItem;
