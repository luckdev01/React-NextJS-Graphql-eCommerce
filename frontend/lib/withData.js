import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include'
        },
        headers
      });
    },
    // local state
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            // read cartOpen from cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY
            });
            // write teh cart state to the opposite
            const data = {
              data: { cartOpen: !cartOpen }
            };
            cache.writeData(data);
            return data;
          }
        }
      },
      defaults: {
        cartOpen: false
      }
    }
  });
}

export default withApollo(createClient);
