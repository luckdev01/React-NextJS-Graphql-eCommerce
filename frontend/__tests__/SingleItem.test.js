import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeItem } from '../lib/testUtils';

describe('<SingleItem />', () => {
  it('renders with proper data', async () => {
    const mocks = [
      {
        // when someone makes a request with this query and variable combo, then return this 'mocked' data
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        result: {
          data: {
            item: fakeItem()
          }
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id={123} />
      </MockedProvider>
    );

    expect(wrapper.text()).toContain('Loading...');
    // this function waits for 0ms, in JS, this places the command on the end of the call stack
    await wait();
    // now wrapper will contain data from query
    wrapper.update();
    // if you do the entire snapchat, the mocked Apollo Query will be mounted as well!
    expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('img'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
  });

  it('Errors with a not found item', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        result: {
          errors: [{ message: 'Items Not Found!' }]
        }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const item = wrapper.find('[data-test="graphql-error"]');
    expect(item.text()).toContain('Items Not Found!');
    expect(toJSON(item)).toMatchSnapshot();
  });
});
