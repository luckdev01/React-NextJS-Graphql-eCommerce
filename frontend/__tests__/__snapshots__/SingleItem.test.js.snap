// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> Errors with a not found item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Items Not Found!
</p>
`;

exports[`<SingleItem /> renders with proper data 1`] = `null`;

exports[`<SingleItem /> renders with proper data 2`] = `null`;

exports[`<SingleItem /> renders with proper data 3`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Network error: No more mocked responses for the query: query SINGLE_ITEM_QUERY($id: ID!) {
  item(where: {id: $id}) {
    id
    title
    description
    largeImage
    __typename
  }
}
, variables: {"id":123}
</p>
`;
